### PR TITLE
Improvements for ruby packages

### PR DESCRIPTION
**Pull request by nforro**

### CODE REVIEW

**Comment by nforro** [comment]
@junaruga: I have implemented `RubyHelper` SPEC hook, so far it is only capable of building source archives using instructions in comments above them. Feel free to test it or comment on the [code](https://github.com/rebase-helper/rebase-helper/blob/25820acedab9f41d58e553387c3a3e5bf43c1e39/rebasehelper/spec_hooks/ruby_helper.py).
I have also added `RubyGems` versioneer for getting versions from `https://rubygems.org`. It is run automatically for ruby packages, with `Anitya` versioneer as a fallback.

**Comment by junaruga** [comment]
@nforro thanks for that.
OK, I will try to test it on your branch `nforro-improvements-for-ruby-packages`, and comment.

**Comment by junaruga** [comment]
I will take a look today from now.
Sorry for my late action.

File: rebasehelper/versioneers/rubygems_versioneer.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    def get_name(cls):
+        return cls.NAME
+
+    @classmethod
+    def get_categories(cls):
+        return cls.CATEGORIES
+
+    @classmethod
+    def _get_version(cls, package_name):
+        # special-case "ruby", as https://rubygems.org/api/v1/gems/ruby.json returns nonsense
```
Comment (junaruga) [inline rebasehelper/versioneers/rubygems_versioneer.py]: Why does this logic exist?
Below URL is reserved by the rubygem web site.
But the package "ruby"' source is not registered in rubygem site.
https://rubygems.org/gems/ruby

Comment (nforro) [reply rebasehelper/versioneers/rubygems_versioneer.py]: Yes, but the API returns some (weird looking) results for it. Just open the [URL](https://rubygems.org/api/v1/gems/ruby.json) and see for yourself.

**Comment by junaruga** [comment]
>  It is run automatically for ruby packages, with Anitya versioneer as a fallback.

There is no rubygems versioneer in the help.
Is it right?

```
(venv) $ rebase-helper --help
...
  --versioneer {anitya,pypi}
                        tool to use for determining latest upstream version
```

File: rebasehelper/versioneers/rubygems_versioneer.py (hunk 1 above)
Comment (junaruga) [reply rebasehelper/versioneers/rubygems_versioneer.py]: OK, I wanted to ask why below logic exists. and that makes sense now.

```
        if package_name == 'ruby':
            return None
```

**Comment by nforro** [comment]
> There is no rubygems versioneer in the help.

You have to run `setup.py build` to regenerate `entry_points` in `egg-info`.

**Comment by junaruga** [comment]
> You have to run setup.py build to regenerate entry_points in egg-info.

Sorry, that's my bad.
I forgot to install rebase-helper on virtualenv.
I was running another path's rebase-helper wrongly.
Now no problem.

```
(venv) $ pip install -e .

(venv) $ pip install --no-cache-dir rpm-py-installer

(venv) $ rebase-helper --help
...
  --versioneer {anitya,pypi,rubygems}
                        tool to use for determining latest upstream version
```

**Comment by junaruga** [comment]
I tested below packages. Other things look good to me.
```
(venv) $ which rebase-helper
/home/jaruga/git/rebase-helper/venv/bin/rebase-helper
```

## ruby

There are 2 points for improvement.

- `2.4.1-None` in the changelog is maybe bug. 
- Is it possible to update not `ruby_version` but `major_version` or `minor_version` or `teeny_version`?


```
(venv) $ pwd
/home/jaruga/fed/ruby

(venv) $ rebase-helper
...
Rebase to ruby-2.4.1.tar.xz SUCCEEDED

(venv) $ diff ruby.spec rebase-helper-results/rebased-sources/ruby.spec

< %global ruby_version %{major_minor_version}.%{teeny_version}
---
> %global ruby_version 2.4.1

> * Mon Sep 04 2017 rebase-helper <rebase-helper@localhost.local> - 2.4.1-None
> - New upstream release 2.4.1
```

## rubygem-bundler

Asking for `/var/tmp/rpm-tmp.uciR35 (%install)` looks bug?

```
(venv) $ pwd
/home/jaruga/fed/rubygem-bundler

(venv) $ rebase-helper
...
Removed files packaged in SPEC file:
/var/tmp/rpm-tmp.uciR35 (%install)
Do you want rebase-helper to try to build the packages one more time [Y/n]? Y
Building SRPM
Building SRPM finished successfully
Building RPMs
Building binary packages failed.
Removed files packaged in SPEC file:
/var/tmp/rpm-tmp.MOFeho (%install)
Do you want rebase-helper to try to build the packages one more time [Y/n]? Y
Running 'text' output tool.
...
Building package failed with unknown reason. Check all available log files.
```

rubygem-childprocess and rubygem-cucumber-core also have same situation with rubygem-bundler.


## rubygem-coffee-script-source

```
(venv) $ pwd
/home/jaruga/fed/rubygem-coffee-script-source

(venv) $ rebase-helper 
...
Building binary packages failed.
We did not find a reason why build failed.
Look at the build log /home/jaruga/git/fedora-packages/rubygem-coffee-script-source/rebase-helper-results/new-build/RPM/build.log
```

Actual situation the build.log => It looks fine `Child return code was: 0`.

but /home/jaruga/git/fedora-packages/rubygem-coffee-script-source/rebase-helper-results/new-build/RPM/root.log
`DEBUG util.py:450:  Error: Some packages could not be found.`

So, is it better to add `root.log` path to the above error message?

## rubygem-byebug

OK

**Comment by nforro** [comment]
Thanks for testing.

> `2.4.1-None` in the changelog is maybe bug.

Yes, definitely. `Release` tag in `ruby.spec` is defined by macro and rebase-helper is not able to deal with that.

> Is it possible to update not `ruby_version` but `major_version`, `minor_version` or `teeny_version` in ruby.spec?

In theory, yes. My plan is to implement some more advanced logic to deal with macros and situations like this. But it will take some time.

> Asking for `/var/tmp/rpm-tmp.uciR35 (%install)` looks bug?

That doesn't seem right. I'll have a look at it.

> So, is it better to add `root.log` path to the above error message?

Yes, that would be better. We could try to find out which of the log files actually contains the failure.

**Comment by junaruga** [comment]
@nforro How is the status?

**Comment by nforro** [comment]
@junaruga: I'm working on the first two issues, it's almost done.

**Comment by junaruga** [comment]
@nforro ok thanks for that. I want to test again after it will be fixed.

By the way, I think that we can split PR.
For example this PR is for first two issue that you are working for. Another PR is for 3rd issue.
Because first two higher priority. 3rd is not higher priority.

**Comment by nforro** [comment]
Yes, those issues aren't actually directly related to this PR or to ruby packages, so I'll open another PR after I'm done.

**Comment by skisela** [comment]
@junaruga after checking the problem with rubygem-bundler and looking at the rebase-helper-results/new-build/RPM/build.log I am getting an error:

`cp: cannot stat '/builddir/build/BUILDROOT/rubygem-bundler-1.15.4-1.fc25.x86_64/usr/share/gems/gems/bundler-1.15.4/lib/bundler/man/gemfile.5': No such file or directory`

even after following the instructions

`# git clone https://github.com/bundler/bundler.git && cd bundler
# git checkout v1.15.4 && tar czvf bundler-1.15.4-specs.tgz spec/`

Can you confirm?

If so, we can agree, that the error is totally misleading and I am already working on the fix.

**Comment by junaruga** [comment]
Hi @skisela , okay, let me confirm it.

**Comment by junaruga** [comment]
@skisela the confirmation is on latest master branch? or `nforro-improvements-for-ruby-packages` branch?

**Comment by skisela** [comment]
@junaruga the latest master branch. Thanks.

**Comment by junaruga** [comment]
@skisela ok.

**Comment by junaruga** [comment]
@skisela when I tried it on latest master branch, I got an another error.
`rebase-helper-results/new-build/RPM/build.log` was not created.

```
(venv) $ pwd
/home/jaruga/fed/rubygem-bundler

(venv) $ rebase-helper -v

...
Traceback (most recent call last):
  File "/home/jaruga/git/rebase-helper/rebasehelper/cli.py", line 268, in run
    app.run()
  File "/home/jaruga/git/rebase-helper/rebasehelper/application.py", line 793, in run
    build = self.build_packages()
  File "/home/jaruga/git/rebase-helper/rebasehelper/application.py", line 526, in build_packages  
    build_dict.update(builder.build(spec, sources, patches, results_dir, **build_dict))
  File "/home/jaruga/git/rebase-helper/rebasehelper/build_helper.py", line 371, in build
    return self._tool.build(*args, **kwargs)
  File "/home/jaruga/git/rebase-helper/rebasehelper/build_tools/mock_tool.py", line 118, in build 
    srpm, cls.logs = cls._build_srpm(spec, sources, patches, results_dir)
  File "/home/jaruga/git/rebase-helper/rebasehelper/build_helper.py", line 288, in _build_srpm
    srpm_results_dir) as tmp_env:
  File "/home/jaruga/git/rebase-helper/rebasehelper/build_helper.py", line 73, in __enter__
    shutil.copy(source, self._env[self.TEMPDIR_SOURCES])
  File "/usr/local/python-3.6.2/lib/python3.6/shutil.py", line 241, in copy
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/local/python-3.6.2/lib/python3.6/shutil.py", line 120, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] No such file or directory: '/home/jaruga/git/fedora-packages/rubygem-bundler/bundler-1.15.4-specs.tgz'
```
```
(venv) $ diff rubygem-bundler.spec ./rebase-helper-results/rebased-sources/rubygem-bundler.spec
12,13c12,13
< Version: 1.13.7
< Release: 3%{?dist}
---
> Version: 1.15.4
> Release: 1%{?dist}
159a160,162
> * Fri Sep 29 2017 rebase-helper <rebase-helper@localhost.local> - 1.15.4-1
> - New upstream release 1.15.4
> 
```

**Comment by junaruga** [comment]
I ran it on clean environment.

```
(venv) $ pwd
/home/jaruga/fed/rubygem-bundler

(venv) $ git clean -fdx

(venv) $ rebase-helper -v
```

**Comment by nforro** [comment]
@junaruga: Actually you have to run it on `nforro-improvements-for-ruby-packages` branch, not `master`, since  there is no `RubyHelper` spec hook there and the source archive is not created. I guess **skisela** didn't realize that.

**Comment by nforro** [comment]
@junaruga: BTW the first two issues from https://github.com/rebase-helper/rebase-helper/pull/327#issuecomment-326939483 related to `ruby` package should be already fixed, so you can test that if you want.

**Comment by junaruga** [comment]
@nforro ok I will try test on `nforro-improvements-for-ruby-packages` branch.

**Comment by junaruga** [comment]
> cp: cannot stat '/builddir/build/BUILDROOT/rubygem-bundler-1.15.4-1.fc25.x86_64/usr/share/gems/gems/bundler-1.15.4/lib/bundler/man/gemfile.5': No such file or directory

I tested for above things about `rubygem-bundler`.
The reason of the error is `gemfile.5` included in Source1 (before), but not in Sourc1 (after).

I suppose we can create new ticket for that.
And I think that this PR is fine.

```
Source1: %{gem_name}-%{version}-specs.tgz
```

```
$ gem fetch -v 1.15.4

$ gem fetch -v 1.13.7

$ gem unpack bundler-1.13.7.gem

$ gem unpack bundler-1.15.4.gem

$ ls bundler-1.13.7/lib/bundler/man/gemfile.5
bundler-1.13.7/lib/bundler/man/gemfile.5

$ ls bundler-1.15.4/lib/bundler/man/gemfile.5
ls: cannot access 'bundler-1.15.4/lib/bundler/man/gemfile.5': No such file or directory
```

**Comment by nforro** [comment]
@junaruga: Thanks, I'm going to merge this. Feel free to report new issues if you find any.

**Pull request merged**